### PR TITLE
Implement dismiss and open behaviors on Menu

### PR DESCRIPTION
**Pull request by rurikoaraki**
### Platforms Impacted
- [ ] iOS
- [x] macOS
- [x] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

Implement open and dismissal behavior for Menu. This PR implements:
- open
- defaultOpen
- onOpenChanged
- openOnHover

This PR also cleans up dismiss behavior. Namely it listens to onDismiss from the Callout now and uses that to change open state (if state is not being controlled)

(this is probably a mess so later me should look over the code again to be sure)

### Verification

Tested cases in Tester.

openOnHover

(mouse moving around is clicking, it's just hard to tell since interaction doesn't show up in the GIF)
openControlled

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/experimental/Menu/src/Menu/Menu.types.ts [hunk 1]
```diff
@@ -1,14 +1,20 @@
...
+import { InteractionEvent } from '@fluentui-react-native/interactive-hooks';
 import React from 'react';
 
 export const menuName = 'Menu';
 
 export interface MenuProps extends Omit<IViewProps, 'onPress'> {
+  defaultOpen?: boolean;
   open?: boolean;
+  onOpenChange?: (e: InteractionEvent, isOpen: boolean) => void;
+  openOnContext?: boolean;
```
Comment (rurikoaraki) [inline packages/experimental/Menu/src/Menu/Menu.types.ts:11]: This should go away for now, not in scope

File: apps/fluent-tester/src/FluentTester/TestComponents/Menu/MenuTest.tsx [hunk 2]
```diff
@@ -57,12 +57,58 @@ const MenuSubMenu: React.FunctionComponent = () => {
...
+            <MenuItem content="A MenuItem" />
+            <Submenu />
+          </MenuList>
+        </MenuPopover>
+      </Menu>
+    </Stack>
+  );
+};
+
+const MenuControlledOpen: React.FunctionComponent = () => {
```
Comment (rurikoaraki) [inline apps/fluent-tester/src/FluentTester/TestComponents/Menu/MenuTest.tsx:78]: Note to self test for pressing the MenuTrigger Button explicitly

File: packages/experimental/Menu/src/MenuPopover/useMenuPopover.ts [hunk 3]
```diff
@@ -1,10 +1,15 @@
...
   const context = useMenuContext();
 
   const triggerRef = context.triggerRef;
+  const onDismiss = context.setOpen(undefined, false /* isOpen */);
+  const dismissBehaviors = context.isControlled
+    ? (['preventDismissOnKeyDown', 'preventDismissOnClickOutside'] as DismissBehaviors[])
+    : undefined;
 
-  return { triggerRef };
+  return { triggerRef, onDismiss, dismissBehaviors };
```
Comment (rurikoaraki) [inline packages/experimental/Menu/src/MenuPopover/useMenuPopover.ts:14]: Type of onDismiss is wrong here, need to fix

**Comment by Saadnajmi** [comment]
Will test shortly on macOS...

File: packages/experimental/Menu/src/MenuPopover/useMenuPopover.ts [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
 
 export const useMenuPopover = (_props: MenuPopoverProps): MenuPopoverState => {
   const context = useMenuContext();
+  const setOpen = context.setOpen;
 
   const triggerRef = context.triggerRef;
+  const onDismiss = React.useCallback(() => setOpen(undefined, false /* isOpen */), [setOpen]);
+  const dismissBehaviors = context.isControlled
+    ? (['preventDismissOnKeyDown', 'preventDismissOnClickOutside'] as DismissBehaviors[])
+    : undefined;
```
Comment (Saadnajmi) [inline packages/experimental/Menu/src/MenuPopover/useMenuPopover.ts:14]: Feels like a good followup in Popover V1 to simplify this bit on the native side..

File: packages/experimental/Menu/src/Menu/useMenu.ts [hunk 5]
```diff
@@ -1,21 +1,44 @@
...
 import React from 'react';
 import { useMenuContext } from '../context/menuContext';
 import { MenuProps, MenuState } from './Menu.types';
 
 export const useMenu = (props: MenuProps): MenuState => {
-  const [open, setOpen] = useMenuOpenState(props);
   const triggerRef = React.useRef(null);
   const context = useMenuContext();
   const isSubmenu = context.triggerRef !== null;
+  const isControlled = typeof props.open === 'undefined' ? false : true;
```
Comment (Saadnajmi) [inline packages/experimental/Menu/src/Menu/useMenu.ts:10]: can we simplify the RHS to `typeof props.open !== 'undefined'` ?

**Review by Saadnajmi — APPROVED** [review]
openOnHover works for macOS, and we don't support controlled Callouts so that doesn't work as expected. Cool stuff!

**Pull request merged**